### DISCUSSION TITLE
Detecting unused variables

### DISCUSSION THREAD

**Discussion by vojtapol** [graphql-java/graphql-java#2412] [Ideas]
GraphQL Java does automatically detect query variables that are unused. That's according to the GraphQL spec. 

For example this query is invalid:

```graphql
query MyQuery($last: Int!) {
  contactConnection(first: 10) {
    edges {
      node {
        id
      }
    }
  }
}
```

What it however does NOT detect are unused variables in the `variables` object:
```graphql
{
  contactConnection(first: 10) {
    edges {
      node {
        id
      }
    }
  }
}
```
```json
{
  "variable": "value"
}
```

Throwing exceptions when extra variables are sent would make detecting some errors easier. For example if a variable is optional in the query, and the developer mistakenly names the variable key differently, it can be hard to find the typo in the variable name.

Unfortunately, `graphql-java` does not support detecting this out of the box. Adding support should be pretty easy though as illustrated by [this Stack Overflow question](https://stackoverflow.com/questions/52834550/graphql-java-throwing-exceptions-when-request-has-unused-variables).

Any thoughts?

**Comment by dugenkui03** [comment]
The 'variables' is the variables defined on the 'operatio', instead of the data value of `ExecutionInput.variables`, [see sepc for details](https://spec.graphql.org/draft/#sec-All-Variables-Used).

Infact, most of validations are data struct validation and data type type validation, instead of data value validation.

**Comment by bbakerman** [comment]
I am torn on this one.  On the one hand it's more precise and it mirrors the spec for operation variables.

But... I personally am not a fan of the UnusedVariablesRule in the spec itself. Because it makes exploratory testing that much harder.

eg Imagine I am in graphiql and making up queries and I used `query op(name : String  age : Int)` and then I decide I dont want to filter by age - I not only have to fix the `$age` usage I also need to fix the operation declaration as well.

So on the one hand once the code gets to production its a good idea but when you are developer mode its a bad idea.  But of course the graphql server can't know its audience.

Anyway I digress with my pet peeve about the graphql spec.

I don't think we should introduce this commonly in graphql-java because it will break everyone out there how is currently sending in more variables than they need (even if they mean to do it ot not) without say a compiler warning like we get on true code breaking changes.

Given that that StackOverflow question is spot on, I wuld suggest that if people want this then can easily add it before they invoke graphql-java code OR write their own Instrumentation that does it on `beforeExecution`

**Reply by vojtapol** [reply]
I understand your view point. For experimentation it can indeed be useful to be lenient and ignore unused variables. If the rule was to be introduced, it would have to be in a backwards compatible manner, and disabled by default.

Nevertheless, I would find this "strict mode" very useful to make the API safer to use. I will try the `beforeExecution` way for now.
